enginetest/table_functions_test.go

```diff
@@ -0,0 +1,190 @@
+// Copyright 2020-2021 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package enginetest_test
```

Comment:
I had a tough time figuring out a clean way to test the base support for table functions at the go-mysql-sever layer. I covered the test cases I wanted to cover, but getting the FunctionProvider set up to return the custom/test table function required relying on the memory engine implementation, so I left it in the `enginetest_test` package because of the coupling with memory engine. I would have preferred it to be in enginetest and reusable by other integrators.

I wasn't super happy with how clean this test code came out, so if you have suggestions to improve, I'd love to hear.

Reply:
I think this is fine. Tests of engine functionality / the memory implementation can already be found in enginetest.go, which is not exported. You might consider putting these tests there, instead of in a new file, but your call.

Another thing you could do is just implement the JSON_TABLE built-in and test that in the main engine tests suite. That gives you reasonable assurance that the mechanism works, and forces you to think about the difference between engine-provided and integrator-provided table functions.

Reply:
Adding JSON_TABLE is a great idea. I'm tempted to jump on that, but I think I should stay focused on getting this change released, instead of adding scope, since I've been working on this for a while.

I'll take your suggestion and move the tests into engine_test.go.